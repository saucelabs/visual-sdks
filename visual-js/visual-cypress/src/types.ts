import {
  DiffingMethod,
  DiffingOptionsIn,
  SauceRegion,
  SelectiveRegionOptions,
} from '@saucelabs/visual';

export interface SauceConfig {
  buildName: string;
  branch?: string;
  defaultBranch?: string;
  project?: string;
  region?: SauceRegion;
  user?: string;
  key?: string;
  diffingMethod?: DiffingMethod;
  diffingOptions?: DiffingOptionsIn;
}

export interface HasSauceConfig {
  saucelabs?: SauceConfig;
}

export type SauceVisualOptions = {
  region: SauceRegion;
};

export type PlainRegion = {
  x: number;
  y: number;
  width: number;
  height: number;
};

export type VisualRegion<
  R extends Omit<object, 'element'> = PlainRegion | Cypress.Chainable,
> = { element: R } & SelectiveRegionOptions;

export type ResolvedVisualRegion = VisualRegion<PlainRegion>;

export type ScreenshotMetadata = {
  id: string;
  name: string;
  testName: string;
  suiteName: string;
  regions: ResolvedVisualRegion[];
  diffingMethod?: DiffingMethod;
  diffingOptions?: DiffingOptionsIn;
  viewport: SauceVisualViewport | undefined;
  devicePixelRatio: number;
  dom?: string;
};

export type SauceVisualViewport = {
  width: number;
  height: number;
};

export type VisualCheckOptions = {
  /**
   * An array of ignore regions or Cypress elements to ignore.
   */
  ignoredRegions?: (PlainRegion | Cypress.Chainable)[];
  /**
   * The diffing method we should use when finding visual changes. Defaults to DiffingMethod.Simple.
   */
  diffingMethod?: DiffingMethod;
  /**
   * The diffing options that should be applied by default.
   */
  diffingOptions?: DiffingOptionsIn;
  /**
   * The diffing method we should use when finding visual changes. Defaults to DiffingMethod.Simple.
   */
  regions?: VisualRegion[];
  /**
   * Enable DOM capture for DOM Inspection and insights.
   */
  captureDom?: boolean;
  /**
   * An HTML selector we should clip to. Can be used for basic component testing
   * / screenshot cropping. Ex: '.class_name', '#id_name', etc
   */
  clipSelector?: string;
  /**
   * Additional options to pass to the Cypress screenshot command.
   */
  cypress?: Partial<
    Cypress.Loggable & Cypress.Timeoutable & Cypress.ScreenshotOptions
  >;
};
