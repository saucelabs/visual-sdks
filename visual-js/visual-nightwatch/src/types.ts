import {
  DiffingMethod,
  FullPageScreenshotOptions,
  getApi,
  RegionIn,
  RegionType,
} from '@saucelabs/visual';
import type {
  Elements,
  NightwatchTestSettingGeneric,
  ScopedElement,
} from 'nightwatch';

declare global {
  var uploadedDiffIds: string[];
  var skipped: number;
}

export type ElementType = ScopedElement | Elements;

export type NightwatchIgnorable =
  | ScopedElement
  | Elements
  | string
  | string[]
  | RegionIn;

export type SauceVisualAPI = ReturnType<typeof getApi>;

export interface CheckOptions {
  ignore?: NightwatchIgnorable[];
  regions?: RegionType<ElementType>[];
  diffingMethod?: DiffingMethod;
  fullPage?: FullPageScreenshotOptions;
  /**
   * Whether we should take a snapshot of the DOM to compare with as a part of the diffing process.
   */
  captureDom?: boolean;
  /**
   * A querySelector compatible selector of an element that we should crop the screenshot to.
   */
  clipSelector?: string;
}

export interface SauceVisualServiceOptions
  extends Pick<CheckOptions, 'fullPage' | 'captureDom' | 'clipSelector'> {
  buildName?: string;
  project?: string;
  branch?: string;
  defaultBranch?: string;
  failOnFailures?: boolean;
}

export type RunnerSettings = NightwatchTestSettingGeneric & {
  sauceVisualService: SauceVisualServiceOptions;
};
