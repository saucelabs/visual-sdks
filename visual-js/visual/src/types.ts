import { RegionIn, SelectorIn } from './graphql/__generated__/graphql';
import { SelectiveRegionOptions } from './common/selective-region';
import { SauceRegion } from './common/regions';

export type FullPageScreenshotOptions<T> =
  | boolean
  | {
      /**
       * @deprecated this field will be removed soon
       */
      addressBarShadowPadding?: number;
      /**
       * Delay in ms after scrolling and before taking screenshots.
       * A slight delay can be helpful if the page is using lazy loading when scrolling
       */
      delayAfterScrollMs?: number;
      /**
       * Disable CSS animations and the input caret in the app.
       */
      disableCSSAnimation?: boolean;
      /**
       * Hide elements on the page after first scroll by css selectors.
       */
      hideAfterFirstScroll?: string[];
      /**
       * Hide all scrollbars in the app.
       */
      hideScrollBars?: boolean;
      /**
       * @deprecated this field will be removed soon
       */
      toolBarShadowPadding?: number;
      /**
       * Limit the number of screenshots taken for scrolling and stitching.
       */
      scrollLimit?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10;
      /**
       * Element used for scrolling (available only in native apps)
       */
      scrollElement?: T | Promise<T>;
      /**
       * XPATH selector of an element that we should crop the screenshot to. Available only on native apps.
       */
      nativeClipSelector?: SelectorIn;
    };

export type Ignorable<T> = T | T[] | Promise<T> | Promise<T[]> | RegionIn;

export type RegionType<T> = { element: T } & SelectiveRegionOptions;

export type ElementMeta = { id: string; name?: string };

export interface VisualEnvOpts {
  user: string | undefined;
  key: string | undefined;
  region: SauceRegion | undefined;
  /**
   * The build ID for the current run.
   */
  buildId: string | null;
  buildName: string;
  project: string | null;
  branch: string | null;
  defaultBranch: string | null;
  customId: string | null;
}
