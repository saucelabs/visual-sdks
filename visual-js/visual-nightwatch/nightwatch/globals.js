const {
  displayStatusTable,
  getApi,
  ensureError,
  BuildMode,
} = require('@saucelabs/visual');
const { buildUrlMessage, validateSauce } = require('../utils/api');
const { VISUAL_BUILD_ID_KEY } = require('../utils/constants');
const SauceVisualCheck = require('./commands/sauceVisualCheck');

let sharedSettings = null;
let sauceConfig = null;
let externalBuildId = false;
const {
  SAUCE_VISUAL_PROJECT,
  SAUCE_VISUAL_BRANCH,
  SAUCE_VISUAL_DEFAULT_BRANCH,
  SAUCE_VISUAL_BUILD_NAME,
  SAUCE_VISUAL_CUSTOM_ID,
} = process.env;

const globals = {
  /**
   * Is being called before all tests are run
   */
  async before(settings) {
    console.log('Sauce Visual service started');
    let buildName,
      project,
      branch,
      defaultBranch,
      visualBuildId,
      visualBuildUrl;
    sharedSettings = settings;

    const {
      sauceVisualService,
      webdriver: { host, port },
    } = sharedSettings;

    //
    // Some error handling
    validateSauce(host, sauceVisualService);

    ({ buildName, project, branch, defaultBranch } = sauceVisualService);
    sauceConfig = {
      hostname: host,
      port: port,
      user: process.env.SAUCE_USERNAME,
      key: process.env.SAUCE_ACCESS_KEY,
    };
    const build = await getExternalBuild(sauceConfig);

    if (build) {
      externalBuildId = true;
      process.env[VISUAL_BUILD_ID_KEY] = build.id;
      ({ url: visualBuildUrl } = build);
    } else {
      const sauceBuildData = {
        name:
          buildName || SAUCE_VISUAL_BUILD_NAME || 'Nightwatch Visual Testing',
        project: project || SAUCE_VISUAL_PROJECT,
        branch: branch || SAUCE_VISUAL_BRANCH,
        defaultBranch: defaultBranch || SAUCE_VISUAL_DEFAULT_BRANCH,
      };
      try {
        ({ id: visualBuildId, url: visualBuildUrl } = await getApi(
          sauceConfig,
        ).createBuild(sauceBuildData));
        process.env[VISUAL_BUILD_ID_KEY] = visualBuildId;
      } catch (e) {
        const errorMessage = ensureError(e).message ?? 'Unknown error';
        console.error(errorMessage);
        throw new Error(errorMessage);
      }
    }
    console.log(buildUrlMessage(visualBuildUrl));
  },

  //
  // The global `beforeEach` and `afterEach` will be invoked before and after each test suite (i.e. test file).
  // This works for both the Default JS/TS and Mocha TestRunners.
  async beforeEach() {
    global.uploadedDiffIds = [];
  },

  /**
   * Is being called after all tests are run
   */
  async after() {
    const visualBuildId = process.env[VISUAL_BUILD_ID_KEY] || '';
    if (!visualBuildId) {
      browser.assert.fail('No buildId found');
      return;
    }
    const {
      sauceVisualService,
      webdriver: { host },
    } = sharedSettings;
    let visualBuildResults, visualBuildUrl;
    const { failOnFailures } = sauceVisualService;

    //
    // Some error handling
    validateSauce(host, sauceVisualService);

    const api = getApi(sauceConfig);

    try {
      if (externalBuildId) {
        ({ url: visualBuildUrl } = await api.build(visualBuildId));
      } else {
        ({ url: visualBuildUrl } = await api.finishBuild({
          uuid: visualBuildId,
        }));
      }
    } catch (e) {
      const errorMessage = ensureError(e).message ?? 'Unknown error';
      browser.assert.fail(errorMessage);
      return;
    }

    if (visualBuildUrl) {
      console.log(buildUrlMessage(visualBuildUrl, { reviewReady: true }));
    }

    try {
      const visualBuild = await api.buildWithDiffs(visualBuildId);
      if (!visualBuild) {
        browser.assert.fail(`build '${visualBuildId}' is unknown`);
        return;
      }

      visualBuildResults = visualBuild.diffs.nodes
        .map((d) => d.status)
        .reduce((r, s) => {
          r[s] = (r[s] ?? 0) + 1;
          return r;
        }, {});
    } catch (e) {
      const errorMessage = ensureError(e).message ?? 'Unknown error';
      browser.assert.fail(errorMessage);
      return;
    }

    displayStatusTable(visualBuildId, visualBuildResults);

    if (failOnFailures) {
      browser.assert.ok(
        visualBuildResults.FAILED === undefined ||
          visualBuildResults.FAILED === 0,
        "The Sauce Visual Service didn't fail.\n",
      );
    }

    console.log('Sauce Visual service stopped\n');
  },

  // The eventBroadcaster is an EventEmitter instance that can be used to listen to various internal Nightwatch events.
  // See https://github.com/nightwatchjs/nightwatch/blob/main/lib/runner/eventHub.js#L62
  registerEventHandlers(eventBroadcaster) {
    //
    // For JS/TS Default to (re)set the `global.uploadedDiffIds` array for each test.
    eventBroadcaster.on('TestRunStarted', (args) => {
      global.uploadedDiffIds = [];
    });

    //
    // For cucumber
    eventBroadcaster.on('TestCaseStarted', (args) => {
      // The beforeEach and TestSuiteStarted are not triggered in CucumberJs
      // that's why we need to (re)set the `global.uploadedDiffIds` array here
      global.uploadedDiffIds = [];
      //
      // Logic to get the feature and scenario name
      const reportData = args.report;
      const testCaseId =
        reportData.testCaseStarted[args.envelope.id].testCaseId;
      const pickleId = reportData.testCases.find(
        (testCase) => testCase.id === testCaseId,
      ).pickleId;
      const pickleData = reportData.pickle.find(
        (pickle) => pickle.id === pickleId,
      );
      const gherkinDocument = reportData?.gherkinDocument.find(
        (document) => document.uri === pickleData.uri,
      );
      const featureData = gherkinDocument.feature;
      const { name: featureName } = featureData;
      const { name: scenarioName } = pickleData;
      SauceVisualCheck.featureName = featureName;
      SauceVisualCheck.scenarioName = scenarioName;
    });
  },
};

const getExternalBuild = async (config) => {
  const buildCompletedMsg = `Sauce Labs Visual: cannot add more screenshots since the build is already completed`;

  if (process.env[VISUAL_BUILD_ID_KEY]) {
    let build;
    try {
      build = await getApi(config).build(process.env[VISUAL_BUILD_ID_KEY]);
    } catch (e) {
      const msg = `Sauce Labs Visual: unable to fetch build for buildId ${process.env[VISUAL_BUILD_ID_KEY]}: ${e}`;
      console.error(msg);
      throw new Error(msg);
    }
    if (build?.mode == BuildMode.Completed) {
      console.error(buildCompletedMsg);
      throw new Error(buildCompletedMsg);
    }
    return build;
  }

  if (SAUCE_VISUAL_CUSTOM_ID) {
    let build;
    try {
      build = await getApi(config).buildByCustomId(SAUCE_VISUAL_CUSTOM_ID);
    } catch (e) {
      const msg = `Sauce Labs Visual: unable to fetch build for customId ${SAUCE_VISUAL_CUSTOM_ID}: ${e}`;
      console.error(msg);
      throw new Error(msg);
    }
    if (build?.mode == BuildMode.Completed) {
      console.error(buildCompletedMsg);
      throw new Error(buildCompletedMsg);
    }
    return build;
  }
};

module.exports = globals;
