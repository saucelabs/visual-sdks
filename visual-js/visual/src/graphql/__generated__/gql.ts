/* eslint-disable */
import * as types from './graphql.js';
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';

/**
 * Map of all GraphQL operations in the project.
 *
 * This map has several performance disadvantages:
 * 1. It is not tree-shakeable, so it will include all operations in the project.
 * 2. It is not minifiable, so the string of a GraphQL query will be multiple times inside the bundle.
 * 3. It does not support dead code elimination, so it will add unused operations.
 *
 * Therefore it is highly recommended to use the babel or swc plugin for production.
 */
const documents = {
    "query build($input: UUID!) {\n  result: build(id: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    defaultBranch\n    mode\n  }\n}": types.BuildDocument,
    "query buildByCustomId($input: String!) {\n  result: buildByCustomId(customId: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    mode\n  }\n}": types.BuildByCustomIdDocument,
    "query buildStatus($input: UUID!) {\n  result: build(id: $input) {\n    url\n    status\n    unapprovedCount: diffCountExtended(input: {status: UNAPPROVED})\n    errorCount: diffCountExtended(input: {status: ERRORED})\n  }\n}": types.BuildStatusDocument,
    "query buildWithDiffs($input: UUID!) {\n  result: build(id: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    diffs {\n      nodes {\n        id\n        baselineId\n        status\n        baseline {\n          snapshot {\n            buildId\n          }\n        }\n        diffBounds {\n          x\n          y\n        }\n        diffClusters {\n          x\n          y\n          width\n          height\n        }\n      }\n    }\n  }\n}": types.BuildWithDiffsDocument,
    "query buildWithDiffsByCustomId($input: String!) {\n  result: buildByCustomId(customId: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    diffs {\n      nodes {\n        id\n        baselineId\n        status\n        baseline {\n          snapshot {\n            buildId\n          }\n        }\n        diffBounds {\n          x\n          y\n        }\n        diffClusters {\n          x\n          y\n        }\n      }\n    }\n  }\n}": types.BuildWithDiffsByCustomIdDocument,
    "mutation createBuild($input: BuildIn!) {\n  result: createBuild(input: $input) {\n    id\n    name\n    project\n    branch\n    defaultBranch\n    status\n    url\n  }\n}": types.CreateBuildDocument,
    "mutation createSnapshot($input: SnapshotIn!) {\n  result: createSnapshot(input: $input) {\n    id\n    uploadId\n    diffs {\n      nodes {\n        id\n        baselineId\n        snapshotId\n        status\n        diffBounds {\n          x\n          y\n          width\n          height\n        }\n        diffClusters {\n          x\n          y\n          width\n          height\n        }\n      }\n    }\n  }\n}": types.CreateSnapshotDocument,
    "mutation createSnapshotFromWebDriver($input: CreateSnapshotFromWebDriverIn!) {\n  result: createSnapshotFromWebDriver(input: $input) {\n    id\n    uploadId\n    diffs {\n      nodes {\n        id\n        baselineId\n        snapshotId\n        status\n        diffBounds {\n          x\n          y\n          width\n          height\n        }\n        diffClusters {\n          x\n          y\n          width\n          height\n        }\n      }\n    }\n  }\n}": types.CreateSnapshotFromWebDriverDocument,
    "mutation createSnapshotUpload($input: SnapshotUploadIn!) {\n  result: createSnapshotUpload(input: $input) {\n    id\n    buildId\n    imageUploadUrl\n    domUploadUrl\n  }\n}": types.CreateSnapshotUploadDocument,
    "query diffsForTestResult($input: UUID!) {\n  result: diffs(condition: {buildId: $input}) {\n    nodes {\n      id\n      status\n    }\n  }\n}": types.DiffsForTestResultDocument,
    "mutation FinishBuildDocument($input: FinishBuildIn!) {\n  result: finishBuild(input: $input) {\n    id\n    name\n    status\n    url\n  }\n}": types.FinishBuildDocumentDocument,
    "mutation mergeBaselines($input: MergeBaselinesInput!) {\n  result: mergeBaselines(input: $input) {\n    baselines {\n      id\n    }\n  }\n}": types.MergeBaselinesDocument,
    "mutation UpdateDiff($input: UpdateDiffIn!) {\n  result: updateDiff(input: $input) {\n    id\n    status\n    baselineId\n    snapshotId\n  }\n}": types.UpdateDiffDocument,
    "query webdriverSessionInfo($input: WebdriverSessionInfoIn!) {\n  result: webdriverSessionInfo(input: $input) {\n    blob\n    operatingSystem\n    operatingSystemVersion\n    browser\n    browserVersion\n    deviceName\n  }\n}": types.WebdriverSessionInfoDocument,
};

/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 *
 *
 * @example
 * ```ts
 * const query = graphql(`query GetUser($id: ID!) { user(id: $id) { name } }`);
 * ```
 *
 * The query argument is unknown!
 * Please regenerate the types.
 */
export function graphql(source: string): unknown;

/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "query build($input: UUID!) {\n  result: build(id: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    defaultBranch\n    mode\n  }\n}"): (typeof documents)["query build($input: UUID!) {\n  result: build(id: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    defaultBranch\n    mode\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "query buildByCustomId($input: String!) {\n  result: buildByCustomId(customId: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    mode\n  }\n}"): (typeof documents)["query buildByCustomId($input: String!) {\n  result: buildByCustomId(customId: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    mode\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "query buildStatus($input: UUID!) {\n  result: build(id: $input) {\n    url\n    status\n    unapprovedCount: diffCountExtended(input: {status: UNAPPROVED})\n    errorCount: diffCountExtended(input: {status: ERRORED})\n  }\n}"): (typeof documents)["query buildStatus($input: UUID!) {\n  result: build(id: $input) {\n    url\n    status\n    unapprovedCount: diffCountExtended(input: {status: UNAPPROVED})\n    errorCount: diffCountExtended(input: {status: ERRORED})\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "query buildWithDiffs($input: UUID!) {\n  result: build(id: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    diffs {\n      nodes {\n        id\n        baselineId\n        status\n        baseline {\n          snapshot {\n            buildId\n          }\n        }\n        diffBounds {\n          x\n          y\n        }\n        diffClusters {\n          x\n          y\n          width\n          height\n        }\n      }\n    }\n  }\n}"): (typeof documents)["query buildWithDiffs($input: UUID!) {\n  result: build(id: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    diffs {\n      nodes {\n        id\n        baselineId\n        status\n        baseline {\n          snapshot {\n            buildId\n          }\n        }\n        diffBounds {\n          x\n          y\n        }\n        diffClusters {\n          x\n          y\n          width\n          height\n        }\n      }\n    }\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "query buildWithDiffsByCustomId($input: String!) {\n  result: buildByCustomId(customId: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    diffs {\n      nodes {\n        id\n        baselineId\n        status\n        baseline {\n          snapshot {\n            buildId\n          }\n        }\n        diffBounds {\n          x\n          y\n        }\n        diffClusters {\n          x\n          y\n        }\n      }\n    }\n  }\n}"): (typeof documents)["query buildWithDiffsByCustomId($input: String!) {\n  result: buildByCustomId(customId: $input) {\n    id\n    name\n    url\n    status\n    project\n    branch\n    diffs {\n      nodes {\n        id\n        baselineId\n        status\n        baseline {\n          snapshot {\n            buildId\n          }\n        }\n        diffBounds {\n          x\n          y\n        }\n        diffClusters {\n          x\n          y\n        }\n      }\n    }\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "mutation createBuild($input: BuildIn!) {\n  result: createBuild(input: $input) {\n    id\n    name\n    project\n    branch\n    defaultBranch\n    status\n    url\n  }\n}"): (typeof documents)["mutation createBuild($input: BuildIn!) {\n  result: createBuild(input: $input) {\n    id\n    name\n    project\n    branch\n    defaultBranch\n    status\n    url\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "mutation createSnapshot($input: SnapshotIn!) {\n  result: createSnapshot(input: $input) {\n    id\n    uploadId\n    diffs {\n      nodes {\n        id\n        baselineId\n        snapshotId\n        status\n        diffBounds {\n          x\n          y\n          width\n          height\n        }\n        diffClusters {\n          x\n          y\n          width\n          height\n        }\n      }\n    }\n  }\n}"): (typeof documents)["mutation createSnapshot($input: SnapshotIn!) {\n  result: createSnapshot(input: $input) {\n    id\n    uploadId\n    diffs {\n      nodes {\n        id\n        baselineId\n        snapshotId\n        status\n        diffBounds {\n          x\n          y\n          width\n          height\n        }\n        diffClusters {\n          x\n          y\n          width\n          height\n        }\n      }\n    }\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "mutation createSnapshotFromWebDriver($input: CreateSnapshotFromWebDriverIn!) {\n  result: createSnapshotFromWebDriver(input: $input) {\n    id\n    uploadId\n    diffs {\n      nodes {\n        id\n        baselineId\n        snapshotId\n        status\n        diffBounds {\n          x\n          y\n          width\n          height\n        }\n        diffClusters {\n          x\n          y\n          width\n          height\n        }\n      }\n    }\n  }\n}"): (typeof documents)["mutation createSnapshotFromWebDriver($input: CreateSnapshotFromWebDriverIn!) {\n  result: createSnapshotFromWebDriver(input: $input) {\n    id\n    uploadId\n    diffs {\n      nodes {\n        id\n        baselineId\n        snapshotId\n        status\n        diffBounds {\n          x\n          y\n          width\n          height\n        }\n        diffClusters {\n          x\n          y\n          width\n          height\n        }\n      }\n    }\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "mutation createSnapshotUpload($input: SnapshotUploadIn!) {\n  result: createSnapshotUpload(input: $input) {\n    id\n    buildId\n    imageUploadUrl\n    domUploadUrl\n  }\n}"): (typeof documents)["mutation createSnapshotUpload($input: SnapshotUploadIn!) {\n  result: createSnapshotUpload(input: $input) {\n    id\n    buildId\n    imageUploadUrl\n    domUploadUrl\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "query diffsForTestResult($input: UUID!) {\n  result: diffs(condition: {buildId: $input}) {\n    nodes {\n      id\n      status\n    }\n  }\n}"): (typeof documents)["query diffsForTestResult($input: UUID!) {\n  result: diffs(condition: {buildId: $input}) {\n    nodes {\n      id\n      status\n    }\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "mutation FinishBuildDocument($input: FinishBuildIn!) {\n  result: finishBuild(input: $input) {\n    id\n    name\n    status\n    url\n  }\n}"): (typeof documents)["mutation FinishBuildDocument($input: FinishBuildIn!) {\n  result: finishBuild(input: $input) {\n    id\n    name\n    status\n    url\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "mutation mergeBaselines($input: MergeBaselinesInput!) {\n  result: mergeBaselines(input: $input) {\n    baselines {\n      id\n    }\n  }\n}"): (typeof documents)["mutation mergeBaselines($input: MergeBaselinesInput!) {\n  result: mergeBaselines(input: $input) {\n    baselines {\n      id\n    }\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "mutation UpdateDiff($input: UpdateDiffIn!) {\n  result: updateDiff(input: $input) {\n    id\n    status\n    baselineId\n    snapshotId\n  }\n}"): (typeof documents)["mutation UpdateDiff($input: UpdateDiffIn!) {\n  result: updateDiff(input: $input) {\n    id\n    status\n    baselineId\n    snapshotId\n  }\n}"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "query webdriverSessionInfo($input: WebdriverSessionInfoIn!) {\n  result: webdriverSessionInfo(input: $input) {\n    blob\n    operatingSystem\n    operatingSystemVersion\n    browser\n    browserVersion\n    deviceName\n  }\n}"): (typeof documents)["query webdriverSessionInfo($input: WebdriverSessionInfoIn!) {\n  result: webdriverSessionInfo(input: $input) {\n    blob\n    operatingSystem\n    operatingSystemVersion\n    browser\n    browserVersion\n    deviceName\n  }\n}"];

export function graphql(source: string) {
  return (documents as any)[source] ?? {};
}

export type DocumentType<TDocumentNode extends DocumentNode<any, any>> = TDocumentNode extends DocumentNode<  infer TType,  any>  ? TType  : never;