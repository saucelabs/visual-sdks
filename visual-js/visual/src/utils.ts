import { FullPageConfigIn } from './graphql/__generated__/graphql';

export const getFullPageConfig: (
  main?: FullPageConfigIn | boolean,
  local?: FullPageConfigIn | boolean,
) => FullPageConfigIn | undefined = (main, local) => {
  const isNoConfig = !main && !local;
  const isLocalOff = local === false;

  if (isNoConfig || isLocalOff) {
    return;
  }

  const globalCfg = typeof main === 'object' ? main : {};
  const localCfg = typeof local === 'object' ? local : {};
  return { ...globalCfg, ...localCfg };
};

export const isSkipMode = (): boolean => {
  const VISUAL_SKIP_ENV_VAR = 'SAUCE_VISUAL_SKIP';
  return Boolean(process.env[VISUAL_SKIP_ENV_VAR]);
};
