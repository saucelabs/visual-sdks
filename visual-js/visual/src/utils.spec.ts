import { describe, expect, test } from '@jest/globals';
import { getFullPageConfig, parseRegionsForAPI } from './utils';
import { RegionIn } from './graphql/__generated__/graphql';
import { FullPageScreenshotOptions } from './types';

type MockElement = { elementId: string };

const configDelay: FullPageScreenshotOptions<MockElement> = {
  delayAfterScrollMs: 1500,
};

const configDelayBig: FullPageScreenshotOptions<MockElement> = {
  delayAfterScrollMs: 5000,
};

/**
 * Simple resolver to test basic string & region parsing.
 * @param itemPromise
 */
const resolveForTest = async (itemPromise: string | Promise<RegionIn>) => {
  const item = await itemPromise;
  if (typeof item === 'object') return [item];
  return [{ id: item }];
};

describe('utils', () => {
  describe('getFullPageConfig', () => {
    describe('returns undefined', () => {
      test('when main is true and local is false', async () => {
        expect(await getFullPageConfig(true, false)).toBeUndefined();
      });
      test('when main is false and local is false', async () => {
        expect(await getFullPageConfig(false, false)).toBeUndefined();
      });
      test('when main is false and local is false', async () => {
        expect(await getFullPageConfig(false, false)).toBeUndefined();
      });
      test('when main is object and local is false', async () => {
        expect(await getFullPageConfig(configDelay, false)).toBeUndefined();
      });
      test('when main is undefined and local is false', async () => {
        expect(await getFullPageConfig(undefined, false)).toBeUndefined();
      });
      test('when main is undefined and local is undefined', async () => {
        expect(await getFullPageConfig(undefined, undefined)).toBeUndefined();
      });
      test('when main is false and local is undefined', async () => {
        expect(await getFullPageConfig(false, undefined)).toBeUndefined();
      });
    });
    describe('returns empty config', () => {
      test('when main is true and local is true', async () => {
        expect(await getFullPageConfig(true, undefined)).toEqual({});
      });
      test('when main is false and local is true', async () => {
        expect(await getFullPageConfig(true, undefined)).toEqual({});
      });
      test('when main is undefined and local is true', async () => {
        expect(await getFullPageConfig(true, undefined)).toEqual({});
      });
      test('when main is true and local is undefined', async () => {
        expect(await getFullPageConfig(true, undefined)).toEqual({});
      });
    });
    describe('returns config', () => {
      test('when main is config and local is true', async () => {
        expect(await getFullPageConfig(configDelay, true)).toEqual(configDelay);
      });
      test('when main is true and local is config', async () => {
        expect(await getFullPageConfig(true, configDelay)).toEqual(configDelay);
      });
      test('when main is false and local is config', async () => {
        expect(await getFullPageConfig(true, configDelay)).toEqual(configDelay);
      });
      test('when main is config and local is config', async () => {
        expect(await getFullPageConfig(configDelay, configDelay)).toEqual(
          configDelay,
        );
      });
      test('and local overwrites main config', async () => {
        expect(await getFullPageConfig(configDelay, configDelayBig)).toEqual(
          configDelayBig,
        );
      });
      test('with merged local and main config', async () => {
        const main = { delayAfterScrollMs: 500 };
        const local = { disableCSSAnimation: false };
        expect(await getFullPageConfig(main, local)).toEqual({
          ...main,
          ...local,
        });
      });
      test('with scrollElement when scrollElement is a promise', async () => {
        const elementId = 'elementId';
        const main = {};
        const local = {
          scrollElement: Promise.resolve({ elementId: elementId }),
        };
        expect(
          await getFullPageConfig(main, local, (el) => el.elementId),
        ).toEqual({
          scrollElement: elementId,
        });
      });
      test('with scrollElement when scrollElement is an object', async () => {
        const elementId = 'elementId';
        const main = { scrollElement: { elementId: elementId } };
        const local = {};
        expect(
          await getFullPageConfig(main, local, (el) => el.elementId),
        ).toEqual({
          scrollElement: elementId,
        });
      });
    });
  });
  describe('parseRegionsForAPI', () => {
    test('parsing string values by the resolver', async () => {
      const result = await parseRegionsForAPI(['test'], resolveForTest);
      expect(result.ignoreRegions).toHaveLength(0);
      expect(result.ignoreElements).toEqual([
        {
          id: 'test',
        },
      ]);
    });

    test('parsing region type values by the resolver', async () => {
      const result = await parseRegionsForAPI<string | RegionIn>(
        [
          {
            element: '1',
            enableOnly: ['content'],
          },
          {
            element: '2',
            disableOnly: ['content'],
          },
          {
            element: { x: 1, y: 1, height: 100, width: 100 },
            disableOnly: ['content'],
          },
        ],
        resolveForTest,
      );
      // Expect our results to be consistent when parsing values
      expect(result.ignoreElements[0]).toMatchSnapshot();
      expect(result.ignoreElements[1]).toMatchSnapshot();
      expect(result.ignoreRegions[0]).toMatchSnapshot();
    });
  });
});
