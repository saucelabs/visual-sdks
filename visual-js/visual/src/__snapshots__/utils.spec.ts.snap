// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils parseRegionsForAPI parsing region type values by the resolver 1`] = `
{
  "diffingOptions": {
    "content": true,
    "dimensions": false,
    "position": false,
    "structure": false,
    "style": false,
    "visual": false,
  },
  "id": "1",
}
`;

exports[`utils parseRegionsForAPI parsing region type values by the resolver 2`] = `
{
  "diffingOptions": {
    "content": false,
    "dimensions": true,
    "position": true,
    "structure": true,
    "style": true,
    "visual": true,
  },
  "id": "2",
}
`;

exports[`utils parseRegionsForAPI parsing region type values by the resolver 3`] = `
{
  "diffingOptions": {
    "content": false,
    "dimensions": true,
    "position": true,
    "structure": true,
    "style": true,
    "visual": true,
  },
  "height": 100,
  "width": 100,
  "x": 1,
  "y": 1,
}
`;
